base/array.jl

```diff
@@ -573,12 +573,11 @@ function collect_to!(dest::AbstractArray{T}, itr, offs, st) where T
     i = offs
     while !done(itr, st)
         el, st = next(itr, st)
-        S = typeof(el)
-        if S === T || S <: T
+        if el isa T || typeof(el) === T
             @inbounds dest[i] = el::T
```

Comment:
if you remove the `typeof(el) === T` condition above, inference will be able to infer the `el::T` result and you can remove this redundant type-check